### ISSUE TITLE
terrainAnalysis(srtmSmooth, bbox) runs  indefinitely

### ISSUE THREAD

**Issue by raulpoppiel** [zecojls/tagee#8]
The function terrainAnalysis from Python package runs indefinitely (or the calculation is too slow compared to the version of the JavaScript code editor) and does not return any result. I think the problem may be in calculateDerivatives.

**Comment by s-kganz** [comment]
Hi @raulpoppiel,

Could you share any more details about what you are trying to do? I just re-ran the example script in the readme and had no issues. Are you calling `.getInfo()` on the result of `terrainAnalysis`? And, how large an area/what resolution are you working in? For large areas, `.getInfo()` is unlikely to succeed and the calculation should be run as an export task.

Thanks!

**Comment by raulpoppiel** [comment]
I encountered an issue while attempting to sample the terrainMetrics using a feature collection (fc) with 4000 points. The visualization works fine, and I can add the terrainMetrics layer to the map without any issues. However, when I try to sample the terrain metrics, the process results in a GEE error: "Error: Computation timed out. (Error code: 3)".

## IT WORKED FINE
`from tagee import terrainAnalysis`

bbox = country_extent.geometry().bounds()
gaussianFilter = ee.Kernel.gaussian(radius=3, sigma=2, units='pixels', normalize=True)
srtmSmooth = ee.Image("USGS/SRTMGL1_003").convolve(gaussianFilter).resample("bilinear")
terrainMetrics = terrainAnalysis(srtmSmooth, bbox)

Map.addLayer(terrainMetrics.select('ShapeIndex'), {}, "Tagee")

## IT RESULTS IN A GEE ERROR
fc = ee.FeatureCollection(about 4000 points)
sampled_points = terrainMetrics.sampleRegions(
    collection=fc,
    scale=30, 
    geometries=False, 
    tileScale=16
)

export_task = ee.batch.Export.table.toDrive(
    collection=sampled_points,
    description='Sampled_Covariates',
    folder=folder,
    fileNamePrefix=file_name,
    fileFormat='CSV'
)
export_task.start()


What am I doing wrong..? Thanks.

**Comment by s-kganz** [comment]
Ah ok, thanks for the context. This could be solved by removing all the clips as part of #7, but there are two other quick fixes you could try until that PR gets merged:

 - Reduce `tileScale`. Since you are sampling over points, few pixels are going into the reducer and a high value for `tileScale` may be splitting up operations unnecessarily and using more computation time. Generally with `tileScale`, running out of *memory* means you should increase it, while running out of *time* means you should decrease it.
 - If messing with `tileScale` doesn't fix it, you could also try splitting your export into chunks. As in, have one task for points 1-1000, a second task for points 1001-2000, etc. I had to do that with my data when I was working with this project a while ago.

Hope that helps, I'll let you know if I find anything else.

**Comment by s-kganz** [comment]
I think your issue is with `tileScale`, the following worked for me (and not incorporating the PR I mentioned above). In my experience high values for `tileScale` are only necessary if you are reducing over a large polygon.

```
# Sample 4k points in Colorado
CO = ee.FeatureCollection("TIGER/2018/States").filter(ee.Filter.eq("NAME", "Colorado")).first()
points = ee.FeatureCollection.randomPoints(CO.geometry(), 4000, 1234)

# Calculate terrain metrics over Colorado
from tagee import terrainAnalysis

gaussianFilter = ee.Kernel.gaussian(
  radius=3, sigma=2, units='pixels', normalize=True
)
srtmSmooth = ee.Image("USGS/SRTMGL1_003").convolve(gaussianFilter).resample("bilinear")

terrainMetrics = terrainAnalysis(srtmSmooth, CO.geometry())

# Sample
sampled_points = terrainMetrics.sampleRegions(
    collection=points,
    scale=30,
    geometries=False,
    tileScale=1
)

# export
export_task = ee.batch.Export.table.toDrive(
    collection=sampled_points,
    description='Sampled_Covariates',
    fileFormat='CSV'
)
export_task.start()
```

**Comment by raulpoppiel** [comment]
Thank you for your assistance, Keenan. After further investigation, I discovered that the issue was not where I initially suspected. I had a combination of datasets including TAGEE, Climate, and Other Data that I was using to sample regions. By running the sampling process individually for each dataset, I identified that a extremely dense dataset (average of >8000 images) was the culprit behind the timeout issue. I resolved this by switching to a a less dense dataset, which has successfully addressed the problem. 

I appreciate your patience and support throughout this process. Thank you once again for your help!
